File: EcoSwapAndroid/src/main/java/com/darrenthiores/ecoswap/android/presentation/add_progress/AddProgressScreen.kt
package com.darrenthiores.ecoswap.android.presentation.add_progress

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.rememberLazyListState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.BottomSheetScaffold
import androidx.compose.material.BottomSheetValue
import androidx.compose.material.ExperimentalMaterialApi
import androidx.compose.material.FabPosition
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Scaffold
import androidx.compose.material.Text
import androidx.compose.material.rememberBottomSheetScaffoldState
import androidx.compose.material.rememberBottomSheetState
import androidx.compose.material.rememberScaffoldState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.input.ImeAction
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.unit.dp
import com.darrenthiores.ecoswap.android.R
import com.darrenthiores.ecoswap.android.presentation.add_progress.components.CarbonActivityDropDown
import com.darrenthiores.ecoswap.android.presentation.add_progress.components.CarbonCategoryDropDown
import com.darrenthiores.ecoswap.android.presentation.add_progress.components.ChallengeSheet
import com.darrenthiores.ecoswap.android.presentation.add_progress.components.ChallengeTaskDropDown
import com.darrenthiores.ecoswap.android.presentation.components.buttons.DefaultButton
import com.darrenthiores.ecoswap.android.presentation.components.buttons.SelectableButton
import com.darrenthiores.ecoswap.android.presentation.components.header.DefaultHeader
import com.darrenthiores.ecoswap.android.presentation.components.textfields.DefaultTextField
import com.darrenthiores.ecoswap.android.theme.SubHeadlineR
import com.darrenthiores.ecoswap.presentation.add_progress.AddProgressEvent
import com.darrenthiores.ecoswap.presentation.add_progress.AddProgressState
import kotlinx.coroutines.launch

@OptIn(ExperimentalMaterialApi::class)
@Composable
fun AddProgressScreen(
    state: AddProgressState,
    onEvent: (AddProgressEvent) -> Unit,
    onBackClick: () -> Unit
) {
    val sheetState = rememberBottomSheetState(initialValue = BottomSheetValue.Collapsed)
    val bottomScaffoldState = rememberBottomSheetScaffoldState(bottomSheetState = sheetState)
    val scaffoldState = rememberScaffoldState()

    val coroutineScope = rememberCoroutineScope()

    val challengeListState = rememberLazyListState()

    val challengeReachEnd = remember(state.challenges.items) {
        derivedStateOf {
            challengeListState.layoutInfo
                .visibleItemsInfo
                .lastOrNull()?.index == state.challenges.items.size -1
        }
    }

    LaunchedEffect(challengeReachEnd.value) {
        if(challengeReachEnd.value && !state.challenges.endReached && !state.challenges.isLoading) {
            onEvent(AddProgressEvent.LoadChallengeNextPage)
        }
    }

    BottomSheetScaffold(
        topBar = {
            DefaultHeader(
                modifier = Modifier
                    .padding(
                        horizontal = 24.dp,
                        vertical = 16.dp
                    ),
                text = stringResource(id = R.string.add_progress),
                onBackClick = onBackClick
            )
        },
        scaffoldState = bottomScaffoldState,
        sheetShape = RoundedCornerShape(topStart = 20.dp, topEnd = 20.dp),
        sheetBackgroundColor = MaterialTheme.colors.surface,
        sheetPeekHeight = 0.dp,
        sheetContent = {
            ChallengeSheet(
                challenges = state.challenges.items,
                onSelectChallenge = {
                    onEvent(AddProgressEvent.OnSelectChallenge(it))

                    coroutineScope.launch {
                        bottomScaffoldState.bottomSheetState.collapse()
                    }
                },
                listState = challengeListState
            )
        }
    ) {
        Scaffold(
            floatingActionButton = {
                DefaultButton(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(
                            horizontal = 24.dp
                        ),
                    label = stringResource(id = R.string.upload),
                    onClick = {
                        onEvent(AddProgressEvent.Upload)
                    },
                    disabled = state.category == null || (
                            state.activity == null && state.challenge == null && state.task == null
                            ) || state.isLoading,
                    isLoading = state.isLoading
                )
            },
            floatingActionButtonPosition = FabPosition.Center,
            scaffoldState = scaffoldState
        ) { paddingValues ->
            LazyColumn(
                modifier = Modifier
                    .fillMaxSize()
                    .padding(paddingValues),
                contentPadding = PaddingValues(
                    top = 16.dp,
                    bottom = 56.dp + 24.dp
                ),
                verticalArrangement = Arrangement.spacedBy(20.dp),
                horizontalAlignment = Alignment.Start
            ) {
                item {
                    Text(
                        modifier = Modifier
                            .padding(
                                horizontal = 24.dp
                            ),
                        text = stringResource(id = R.string.category_add),
                        style = SubHeadlineR
                    )

                    Spacer(modifier = Modifier.height(8.dp))

                    CarbonCategoryDropDown(
                        modifier = Modifier
                            .padding(
                                horizontal = 24.dp
                            ),
                        category = state.category,
                        isOpen = state.categoryDropDownOpen,
                        onClick = {
                            onEvent(AddProgressEvent.ToggleCategoryDropDown)
                        },
                        onDismiss = {
                            onEvent(AddProgressEvent.ToggleCategoryDropDown)
                        },
                        onSelectCategory = {
                            onEvent(AddProgressEvent.OnSelectCategory(it))
                        }
                    )
                }

                when(state.category?.id) {
                    "1" -> {
                        item {
                            Text(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = stringResource(id = R.string.activity),
                                style = SubHeadlineR
                            )

                            Spacer(modifier = Modifier.height(8.dp))

                            CarbonActivityDropDown(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                activities = state.activities,
                                activity = state.activity,
                                isOpen = state.activityDropDownOpen,
                                onClick = {
                                    onEvent(AddProgressEvent.ToggleActivityDropDown)
                                },
                                onDismiss = {
                                    onEvent(AddProgressEvent.ToggleActivityDropDown)
                                },
                                onSelectActivity = {
                                    onEvent(AddProgressEvent.OnSelectActivity(it))
                                }
                            )
                        }

                        item {
                            Text(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = stringResource(id = R.string.meter),
                                style = SubHeadlineR
                            )

                            Spacer(modifier = Modifier.height(8.dp))

                            DefaultTextField(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = state.number,
                                onTextChange = {
                                    onEvent(AddProgressEvent.OnNumberChange(it))
                                },
                                placeholder = stringResource(id = R.string.transport_placeholder),
                                keyboardOptions = KeyboardOptions(
                                    imeAction = ImeAction.Done,
                                    keyboardType = KeyboardType.NumberPassword
                                )
                            )
                        }
                    }
                    "2" -> {
                        item {
                            Text(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = stringResource(id = R.string.activity),
                                style = SubHeadlineR
                            )

                            Spacer(modifier = Modifier.height(8.dp))

                            CarbonActivityDropDown(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                activities = state.activities,
                                activity = state.activity,
                                isOpen = state.activityDropDownOpen,
                                onClick = {
                                    onEvent(AddProgressEvent.ToggleActivityDropDown)
                                },
                                onDismiss = {
                                    onEvent(AddProgressEvent.ToggleActivityDropDown)
                                },
                                onSelectActivity = {
                                    onEvent(AddProgressEvent.OnSelectActivity(it))
                                }
                            )
                        }

                        item {
                            Text(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = stringResource(id = R.string.minute),
                                style = SubHeadlineR
                            )

                            Spacer(modifier = Modifier.height(8.dp))

                            DefaultTextField(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = state.number,
                                onTextChange = {
                                    onEvent(AddProgressEvent.OnNumberChange(it))
                                },
                                placeholder = stringResource(id = R.string.energy_placeholder),
                                keyboardOptions = KeyboardOptions(
                                    imeAction = ImeAction.Done,
                                    keyboardType = KeyboardType.NumberPassword
                                )
                            )
                        }
                    }
                    "3" -> {
                        item {
                            Text(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = stringResource(id = R.string.challenge),
                                style = SubHeadlineR
                            )

                            Spacer(modifier = Modifier.height(8.dp))

                            SelectableButton(
                                modifier = Modifier
                                    .padding(
                                        horizontal = 24.dp
                                    ),
                                text = stringResource(id = R.string.select_challenge),
                                value = state.challenge?.title,
                                onClick = {
                                    coroutineScope.launch {
                                        bottomScaffoldState.bottomSheetState.expand()
                                    }
                                }
                            )
                        }

                        state.challenge?.let { challenge ->
                            item {
                                Text(
                                    modifier = Modifier
                                        .padding(
                                            horizontal = 24.dp
                                        ),
                                    text = stringResource(id = R.string.task),
                                    style = SubHeadlineR
                                )

                                Spacer(modifier = Modifier.height(8.dp))

                                ChallengeTaskDropDown(
                                    modifier = Modifier
                                        .padding(
                                            horizontal = 24.dp
                                        ),
                                    tasks = challenge.tasks,
                                    task = state.task,
                                    isOpen = state.taskDropDownOpen,
                                    onClick = {
                                        onEvent(AddProgressEvent.ToggleTaskDropDown)
                                    },
                                    onDismiss = {
                                        onEvent(AddProgressEvent.ToggleTaskDropDown)
                                    },
                                    onSelectTask = {
                                        onEvent(AddProgressEvent.OnSelectTask(it))
                                    }
                                )
                            }
                        }
                    }
                }
            }
        }
    }
}

File: EcoSwapiOS/EcoSwapiOS/Presentation/AddProgress/AddProgressScreen.swift
//
//  AddProgressScreen.swift
//  EcoSwapiOS
//
//  Created by Darren Thiores on 11/01/24.
//  Copyright © 2024 orgName. All rights reserved.
//

import SwiftUI
import shared

struct AddProgressScreen: View {
    var challengeId: String? = nil
    @Binding var shouldRefresh: Bool
    
    @Environment(\.dismiss) private var dismiss
    @StateObject private var viewModel = IosAddProgressViewModel()
    @State private var showSheet: Bool = false
    
    var body: some View {
        ScrollView (.vertical, showsIndicators: false){
            LazyVStack(alignment: .leading) {
                Spacer()
                    .frame(height: 16)
                
                Group {
                    Text("Category*")
                        .font(.SubHeadlineR)
                    
                    Spacer()
                        .frame(height: 8)
                    
                    CarbonCategoryDropDown(
                        category: viewModel.state.category,
                        isOpen: viewModel.state.categoryDropDownOpen,
                        selectCategory: { category in
                            viewModel.onEvent(
                                event: .OnSelectCategory(category: category)
                            )
                        },
                        toggleIsOpen: {
                            viewModel.onEvent(
                                event: .ToggleCategoryDropDown()
                            )
                        }
                    )
                    
                    Spacer()
                        .frame(height: 20)
                }
                .padding(.horizontal, 24)
                
                if viewModel.state.category?.id == "1" || viewModel.state.category?.id == "2" {
                    Group {
                        Text("Activity*")
                            .font(.SubHeadlineR)
                        
                        Spacer()
                            .frame(height: 8)
                        
                        CarbonActivityDropDown(
                            activities: viewModel.state.activities,
                            activity: viewModel.state.activity,
                            isOpen: viewModel.state.activityDropDownOpen,
                            selectActivity: { activity in
                                viewModel.onEvent(
                                    event: .OnSelectActivity(activity: activity)
                                )
                            },
                            toggleIsOpen: {
                                viewModel.onEvent(
                                    event: .ToggleActivityDropDown()
                                )
                            }
                        )
                        
                        Spacer()
                            .frame(height: 20)
                    }
                    .padding(.horizontal, 24)
                    
                    Group {
                        Text(
                            (viewModel.state.category?.id == "1") ? "Number (in Meter)*" : "Number (in Minute)*"
                        )
                        .font(.SubHeadlineR)
                        
                        Spacer()
                            .frame(height: 8)
                        
                        DefaultTextField(
                            text: Binding(
                                get: {
                                    viewModel.state.number
                                },
                                set: {
                                    viewModel.onEvent(
                                        event: .OnNumberChange(number: $0)
                                    )
                                }
                            ),
                            placeholder: (viewModel.state.category?.id == "1") ? "How far you do the task (in Meter)" : "How long you do the task (in Minute)"
                        )
                        .keyboardType(.numberPad)
                        
                        Spacer()
                            .frame(height: 20)
                    }
                    .padding(.horizontal, 24)
                }
                
                if viewModel.state.category?.id == "3" {
                    Group {
                        Text("Challenge*")
                            .font(.SubHeadlineR)
                        
                        Spacer()
                            .frame(height: 8)
                        
                        SelectableButton(
                            text: "Select Challenge",
                            value: viewModel.state.challenge?.title,
                            onClick: {
                                showSheet = true
                            }
                        )
                        
                        Spacer()
                            .frame(height: 20)
                    }
                    .padding(.horizontal, 24)
                    
                    if let challenge = viewModel.state.challenge {
                        Group {
                            Text("Task*")
                                .font(.SubHeadlineR)
                            
                            Spacer()
                                .frame(height: 8)
                            
                            ChallengeTaskDropDown(
                                tasks: challenge.tasks,
                                task: viewModel.state.task,
                                isOpen: viewModel.state.taskDropDownOpen,
                                selectTask: { task in
                                    viewModel.onEvent(
                                        event: .OnSelectTask(task: task)
                                    )
                                },
                                toggleIsOpen: {
                                    viewModel.onEvent(
                                        event: .ToggleTaskDropDown()
                                    )
                                }
                            )
                            
                            Spacer()
                                .frame(height: 20)
                        }
                        .padding(.horizontal, 24)
                    }
                }
                
                Spacer()
                    .frame(height: 24)
            }
        }
        .onAppear {
            viewModel.startObserving()
            
            if let challengeId = challengeId {
                viewModel.onEvent(
                    event: .InitByChallenge(
                        challengeId: challengeId
                    )
                )
            }
        }
        .onDisappear {
            viewModel.dispose()
        }
        .navigationBarBackButtonHidden()
        .toolbar(.hidden, for: .tabBar)
        .navigationBarTitleDisplayMode(.inline)
        .toolbar {
            ToolbarItem(placement: .principal) {
                DefaultHeader(
                    title: "Add Progress",
                    onBackClick: {
                        dismiss()
                    }
                )
                .padding(16)
            }
            
            ToolbarItem(placement: .bottomBar) {
                ButtonFill(
                    label: "Upload",
                    onClick: {
                        viewModel.onEvent(
                            event: .Upload()
                        )
                    },
                    disabled: viewModel.state.category == nil ||
                    (viewModel.state.activity == nil && viewModel.state.challenge == nil) || (viewModel.state.challenge != nil && viewModel.state.task == nil) || (viewModel.state.activity != nil && viewModel.state.number.isEmpty) || viewModel.state.isLoading,
                    fillWidth: true,
                    isLoading: viewModel.state.isLoading
                )
                .padding(.horizontal, 24)
                .padding(.vertical, 8)
            }
        }
        .sheet(isPresented: $showSheet) {
            ChallengeSheet(
                challenges: viewModel.state.challenges.items as? [Challenge] ?? [],
                onSelectChallenge: { challenge in
                    viewModel.onEvent(
                        event: .OnSelectChallenge(challenge: challenge)
                    )
                    
                    showSheet = false
                },
                onAppear: {
                    if !viewModel.state.challenges.endReached && !viewModel.state.challenges.isLoading {
                        viewModel.onEvent(event: .LoadChallengeNextPage())
                    }
                }
            )
        }
        .onChange(of: viewModel.state.isSuccess) { success in
            if success {
                shouldRefresh = true
                showSheet = false
                dismiss()
            }
        }
    }
}

#Preview {
    AddProgressScreen(
        challengeId: nil,
        shouldRefresh: .constant(false)
    )
}
